### PR TITLE
feat: W4 Orchestrator + Verification + PR Control (#4)

### LINKED ISSUE #4: [Epic] W4: Orchestrator + Verification + PR Control
## Overview

Build the orchestration engine that assigns tickets to agents, runs them autonomously, verifies their work through QA, and controls PR creation.

### Orchestrator
- Execution queue with concurrency control
- Ticket-to-agent assignment based on category/role matching
- Branch creation per ticket (feature/TICKET-ID)
- Parallel agent execution with lane-based isolation

### Self-Verification Loop
- QA Agent verification pipeline: test execution + lint + type check
- Convention compliance checking against CONVENTIONS.md
- LLM-based code review (Thinking: LOW for efficiency)
- Notepad system: learnings.md, issues.md, verification.md
- Retry logic: feedback to original agent + auto-retry (max N times)
- Escalation to user when max retries exceeded
- Verification result recording to database

### PR Control System
- Manual mode: all PRs require user review
- Smart mode (default): rule-based auto/manual classification
- Auto mode: QA + Customer PASS → auto merge
- Auto-merge rules engine: maxFilesChanged, forbiddenPaths, requiredTests, etc.
- PR creation with ticket info + verification results in body
- PR template generation

### From oh-my-opencode
- **Atlas verification protocol**: Never trust subagent claims; run lsp_diagnostics, full test suite, read actual files
- **BackgroundManager state machine**: 6-state model (pending → running → completed/error/cancelled/stale)
- **Concurrency slots**: Per-provider/per-model limits prevent rate-limit cascades
- **Stability detection**: 3 consecutive unchanged polls before marking completed

### From OpenClaw
- **Lane-based concurrency**: Serial per-session by default, parallel across sessions
- **Queue modes**: collect, steer, followup for different message handling strategies
- **Default Serial, Explicit Parallel**: Prevents state corruption

## Success Criteria
- [ ] Orchestrator assigns tickets to correct agents based on role
- [ ] Agents execute tickets autonomously with branch isolation
- [ ] QA verification catches test failures and lint errors
- [ ] Convention violations detected and reported
- [ ] Failed verification triggers retry with feedback
- [ ] PR created with correct metadata after verification passes
- [ ] Smart mode correctly classifies PRs as auto/manual

## Dependencies
- #3 W3: Data Layer + Goal Decomposition
- #2 W2: Agent System + Tool Layer

## Estimated Effort
~5 days (1 developer)

### PR DESCRIPTION
**Pull request by 12OneTwo12**
## Summary

Implements [Issue #4](https://github.com/12OneTwo12/phalanx/issues/4) — W4: Orchestrator + Verification + PR Control.

Closes #4

### Agent Ticket Executor + Branch Isolation
- **`AgentTicketExecutor`** — `TicketExecutor` implementation that creates isolated branches, resolves agent config from ticket metadata, builds context-rich prompts, delegates to `AgentExecutor`
- **`BranchManager`** — Git branch lifecycle management with `ticket/{id}-{slug}` naming, slug normalization, `GitOperations` abstraction for testability

### QA Verification Pipeline (`engine/verification/`)
- **`VerificationCheck`** interface (Strategy pattern) + 5 built-in checks:
  - `TestRunnerCheck` — runs project test suite
  - `LintCheckerCheck` — runs linter
  - `TypeCheckerCheck` — runs `tsc --noEmit`
  - `ConventionCheckerCheck` — validates against CONVENTIONS.md
  - `LLMCodeReviewCheck` — LLM-based code review (thinking: LOW)
- **`QAVerificationStrategy`** — `VerificationStrategy` implementation running all checks sequentially
- **`Notepad`** — per-ticket file-based learning/issue/verification records

### PR Control System (`engine/pr/`)
- **`PRController`** — 3 modes: Manual (always manual), Smart (rule-based), Auto (always auto-merge)
- **`SmartModeEngine`** — 5 built-in `AutoMergeRule`s (OCP-compliant):
  - MaxFilesChangedRule, ForbiddenPathsRule, RequireTestsPassRule, ForbiddenKeywordsRule, NoNewDependenciesRule
- **`PRTemplate`** — generates PR body with ticket info, verification results, changed files
- **`PRCreator`** — assembles and creates PRs

### Ticket Pipeline
- **`TicketPipeline`** — E2E facade: branch creation → agent execution → QA verification → PR creation
- Stability detection, retry logic, event-based progress tracking

## Design Principles
- **SOLID** — Strategy (VerificationCheck, AutoMergeRule), Facade (TicketPipeline), DI throughout
- **Dependency direction**: `pipeline → orchestrator → executor → agent`, `pipeline → verification → checks`, `pipeline → pr-controller → pr-creator`
- All external deps (git, shell, LLM) abstracted behind interfaces
- Zero `any` in source code

## Quality Assurance
- 3 review rounds:
  - Round 1: Fixed TicketPipeline ISP violation (VerificationService → VerificationStrategy)
  - Round 2: Added 30 edge-case tests (branch edge cases, concurrent failures, rule boundaries)
  - Round 3: Full completeness audit — all 14 scope items verified as real implementations

## Tests
- **521 tests across 52 files — all passing** ✅
- 104 new tests (74 initial + 30 edge cases)

### CODE REVIEW

File: packages/core/src/engine/ticket-pipeline.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+          epicId: ticket.epicId,
+        };
+
+        const prResult = await this.prCreator.create({
+          ticket: ticketInfo,
+          branch: ticket.branch ?? '',
+          baseBranch: this.config.baseBranch,
+          verificationResult,
+          decision: prDecision.decision,
+        });
```
[the remark this answers is not in the document]
Comment (12OneTwo12) [reply packages/core/src/engine/ticket-pipeline.ts:162]: Fixed in `bcb6b84`. The pipeline now calls `branchManager.getCurrentBranch()` to get the actual branch name instead of relying on `ticket.branch` which may not be updated after branch creation. Falls back to `ticket.branch ?? ""` if getCurrentBranch fails.

File: packages/core/src/engine/pr/smart-mode-rules.ts [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+  forbiddenKeywords?: string[];
+}): SmartModeEngine {
+  return new SmartModeEngine([
+    new MaxFilesChangedRule(config?.maxFiles),
+    new ForbiddenPathsRule(config?.forbiddenPaths),
+    new RequireTestsPassRule(),
+    new ForbiddenKeywordsRule(config?.forbiddenKeywords),
+    new NoNewDependenciesRule(),
+  ]);
+}
```
[the remark this answers is not in the document]
Comment (12OneTwo12) [reply packages/core/src/engine/pr/smart-mode-rules.ts]: Fixed in `bcb6b84`. `createDefaultSmartModeEngine` now imports and uses `DEFAULT_AUTO_MERGE_RULE_CONFIG` as the baseline, merging with any user-provided partial config. The factory parameter type is now `Partial<AutoMergeRuleConfig>` for consistency. All built-in rules (MaxFilesChanged, ForbiddenPaths, ForbiddenKeywords) now receive values from the merged config.

File: packages/core/src/engine/agent-ticket-executor.ts [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+   * 3. Running the agent
+   * 4. Converting result to TicketExecutor format
+   */
+  async execute(ticket: Ticket): Promise<{ success: boolean; error?: string }> {
+    const slug = this.slugify(ticket.title);
+    let branchName: string | undefined;
+
+    try {
+      // Create isolated branch for this ticket
+      branchName = await this.branchManager.createTicketBranch(ticket.id, slug);
```
[the remark this answers is not in the document]
Comment (12OneTwo12) [reply packages/core/src/engine/agent-ticket-executor.ts]: Fixed in `bcb6b84`. Removed the private `slugify` method from `AgentTicketExecutor`. Now passes the raw ticket title directly to `BranchManager.createTicketBranch()`, which handles slug normalization via its own `normalizeSlug()` — single source of truth for branch-safe slug creation.
